quinn-proto/src/cid_queue.rs

```diff
@@ -31,60 +31,83 @@ impl CidQueue {
         }
     }
 
-    pub fn insert(&mut self, cid: IssuedCid) -> Result<(), InsertError> {
-        if cid.sequence == self.offset && self.buffer[self.cursor].is_some() {
-            // This is a duplicate of the active CID.
-            return Ok(());
-        }
-        let index = cid
-            .sequence
-            .checked_sub(self.offset)
-            .ok_or(InsertError::Retired)?;
-        if index >= Self::LEN as u64 {
+    /// Handle a `NEW_CONNECTION_ID` frame
+    ///
+    /// Returns a non-empty range of retired sequence numbers and the reset token of the new active
+    /// CID iff any CIDs were retired.
+    pub fn insert(
+        &mut self,
+        cid: NewConnectionId,
+    ) -> Result<Option<(Range<u64>, ResetToken)>, InsertError> {
+        // Position of new CID wrt. the current active CID
+        let index = match cid.sequence.checked_sub(self.offset) {
+            None => return Err(InsertError::Retired),
+            Some(x) => x,
+        };
+
+        let retired_count = cid.retire_prior_to.saturating_sub(self.offset);
```

Comment:
If the list had gaps, is this still the right number? I guess it doesn't matter since it resets anything to `None` anyway

Reply:
CIDs must be issued in order, so semantically the gaps don't matter, and we take special care not to do work for CIDs between initial offset + LEN and the first unretired CID. It's true that `retired_count` is not the number of actually known CIDs that are retired, but that number isn't what we need anywhere anyway.